From 47deee59a8b1d192ae528fb16bb34986f2703f33 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Joachim=20Mairb=C3=B6ck?= <j.mairboeck@gmail.com>
Date: Wed, 2 Sep 2026 20:38:13 +0200
Subject: adjust paths for Haiku, add executable_path implementation

The fallback based on the environment PATH doesn't work, because that
is just /bin and not suitable for constructing a relative path from it.

diff --git a/src/engine/pathsys.cpp b/src/engine/pathsys.cpp
index 88b0ee1..2ecc633 100644
--- a/src/engine/pathsys.cpp
+++ b/src/engine/pathsys.cpp
@@ -449,6 +449,19 @@ char * executable_path( char const * argv0 )
 
     return posix_path > 0 ? strdup( posix_path ) : NULL;
 }
+#elif defined(__HAIKU__)
+# include <image.h>
+char * executable_path( char const * argv0 )
+{
+    int32 cookie = 0;
+    image_info info;
+    while ( get_next_image_info( B_CURRENT_TEAM, &cookie, &info ) == B_OK )
+    {
+        if ( info.type == B_APP_IMAGE )
+            return strndup( info.name, MAXPATHLEN );
+    }
+    return NULL;
+}
 #else
 char * executable_path( char const * argv0 )
 {
diff --git a/src/engine/startup.cpp b/src/engine/startup.cpp
index 4ec3993..246a7a1 100644
--- a/src/engine/startup.cpp
+++ b/src/engine/startup.cpp
@@ -182,9 +182,9 @@ bool b2::startup::bootstrap(FRAME * frame)
 								// install location.
 			".b2/",
 			// Check relative to the exec for system install location.
-			"../share/b2/",
+			"../data/b2/",
 			// Check relative to the exec for legacy install location.
-			"../share/b2/src/",
+			"../data/b2/src/",
 			// Check development location relative to executable in src/engine.
 			"../",
 			// Check development location relative to executable at root.
diff --git a/src/tools/package.jam b/src/tools/package.jam
index 001206b..6459340 100644
--- a/src/tools/package.jam
+++ b/src/tools/package.jam
@@ -72,8 +72,8 @@ class package-paths
         handle-path prefix : $(default-prefix) ;
         handle-path libdir : $(self.prefix)/lib ;
         handle-path bindir : $(self.prefix)/bin ;
-        handle-path includedir : $(self.prefix)/include ;
-        handle-path datarootdir : $(self.prefix)/share ;
+        handle-path includedir : $(self.prefix)/develop/headers ;
+        handle-path datarootdir : $(self.prefix)/data ;
     }
 
     local rule handle-path ( option : default-value )
diff --git a/src/tools/stage.jam b/src/tools/stage.jam
index 89356d6..9da9899 100644
--- a/src/tools/stage.jam
+++ b/src/tools/stage.jam
@@ -37,9 +37,9 @@ args.add-arg bindir : --bindir
 args.add-arg libdir : --libdir
     : "Install libraries to this directory. [default: <prefix>/lib]" ;
 args.add-arg includedir : --includedir
-    : "install include files to this directory. [default: <prefix>/include]" ;
+    : "install include files to this directory. [default: <prefix>/develop/headers]" ;
 args.add-arg datarootdir : --datarootdir
-    : "Install shared data files to this directory. [default: <prefix>/share]" ;
+    : "Install shared data files to this directory. [default: <prefix>/data]" ;
 
 
 feature.feature <install-dependencies> : off on : incidental      ;
-- 
2.54.0

